[TensorPipe] Update documentation

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40222 [TensorPipe] Update documentation**

Mention the TensorPipe agent in the RPC docs and give users the information they need to choose which agent to use.

Differential Revision: [D22116494](https://our.internmc.facebook.com/intern/diff/D22116494/)

lw committed:
[TensorPipe] Update documentation

Mention the TensorPipe agent in the RPC docs and give users the information they need to choose which agent to use.

Differential Revision: [D22116494](https://our.internmc.facebook.com/intern/diff/D22116494/)

[ghstack-poisoned]

mrshenli commented on docs/source/rpc.rst:
```diff
@@ -122,8 +116,82 @@ how a given function should be treated on the callee side.
+
+
+Process Group Backend
+^^^^^^^^^^^^^^^^^^
```
missing three "^^^"

lw replied:
Fixed

mrshenli commented on docs/source/rpc.rst:
```diff
@@ -122,8 +116,82 @@ how a given function should be treated on the callee side.
 
+.. _rpc-backends:
+
+RPC Backends
```
Shall we make this as a subsection in "RPC"?

lw replied:
Done

mrshenli commented on docs/source/rpc.rst:
```diff
@@ -122,8 +116,82 @@ how a given function should be treated on the callee side.
+^^^^^^^^^^^^^^^^^^
+
+The TensorPipe agent leverages `the TensorPipe
+library <https://github.com/pytorch/tensorpipe>`_, which provides a natively
```
natively provides or just native?

lw replied:
I meant it as "it is point-to-point in a native way", rather than being collective with point-to-point being an "abuse" of the API.

mrshenli commented on docs/source/rpc.rst:
```diff
@@ -122,8 +116,82 @@ how a given function should be treated on the callee side.
+
+TensorPipe Backend
+^^^^^^^^^^^^^^^^^^
+
```
Let's add a warning to mention that this is experimental?

lw replied:
Done (though I think the terminology now is beta, right?)

mrshenli reviewed (commented):
Thank you very much for updating docs!!

lw committed:
Update on "[TensorPipe] Update documentation"


Mention the TensorPipe agent in the RPC docs and give users the information they need to choose which agent to use.

Differential Revision: [D22116494](https://our.internmc.facebook.com/intern/diff/D22116494/)

[ghstack-poisoned]

lw commented:
I compiled it and looked at it and fixed a few issues. I added an example of how to enable TensorPipe. I added a warning that it is beta.

Also, I may have been a bit biased towards TensorPipe when comparing it to ProcessGroup, let me know if you want to counterbalance.

lw committed:
Update on "[TensorPipe] Update documentation"


Mention the TensorPipe agent in the RPC docs and give users the information they need to choose which agent to use.

Differential Revision: [D22116494](https://our.internmc.facebook.com/intern/diff/D22116494/)

[ghstack-poisoned]

lw commented:
Rebased and fixed merge conflict

lw commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -473,7 +454,38 @@ PyObject* rpc_init(PyObject* /* unused */) {
+                  thread-pool used by
+                  :class:`~torch.distributed.rpc.TensorPipeAgent` to execute
+                  requests (default: 16).
+              _transports (list, optional): THIS IS AN ADVANCED PARAMETER, DO
+                  NOT USE UNLESS INSTRUCTED TO. The names of the TensorPipe
+                  transports that the agent should register, in order of
+                  decreasing priority.
+              _channels (list, optional): THIS IS AN ADVANCED PARAMETER, DO NOT
+                  USE UNLESS INSTRUCTED TO. The names of the TensorPipe channels
+                  that the agent should register, in order of decreasing
+                  priority.
```
I was wondering: should we even include these, or should we hide them? It may make sense to hide them, because in the expected scenario (i.e., no transports/channels is defective) we don't want anyone to use them.

mrshenli replied:
+1 to hide them, if users really need it, we can then tell them there is such args here.

mrshenli reviewed (approved):
Thanks for adding docs!

lw committed:
Update on "[TensorPipe] Update documentation"


Mention the TensorPipe agent in the RPC docs and give users the information they need to choose which agent to use.

Differential Revision: [D22116494](https://our.internmc.facebook.com/intern/diff/D22116494/)

[ghstack-poisoned]
